File: 백준/Gold/3079. 입국심사/입국심사.py
import sys
input = sys.stdin.readline

N, M = map(int, input().split())
T = [int(input()) for _ in range(N)]
T.sort()

l = 1
r = 10**20
ans = r
while l <= r:
    mid = (l+r)//2
    cnt = 0
    for t in T:
        if t > mid:
            break
        cnt += mid//t
    if cnt >= M:
        ans = min(ans, mid)
        r = mid - 1
    else:
        l = mid + 1
print(ans)

File: 프로그래머스/3/43238. 입국심사/입국심사.cs
using System;

public class Solution {
    long ANS;
    int[] TIME;
    int N;
    
    public bool check(long mid) {
        long cnt = 0;
        for (int i=0; i<TIME.Length; ++i) {
            cnt += mid/TIME[i];
        }
        if (cnt>=N) return true;
        return false;
    }
    
    
    public long solution(int n, int[] times) {
        ANS = 0;
        TIME = times;
        N = n;
        
        long l = 0;
        long r = 1_000_000_000_000_000_000L;
        while (l <= r) {
            long mid = (l+r) >> 1;
            if (check(mid)) {
                // 가능
                ANS = mid;
                r = mid-1;
            }    
            else {
                // 불가능
                l = mid+1;
            }
        }
        
        return ANS;
    }
}